File: backend/src/main/java/br/ufpr/sept/so2/modules/bff/api/dto/AlunoDashboardResponse.java
package br.ufpr.sept.so2.modules.bff.api.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@JsonInclude(JsonInclude.Include.ALWAYS)
public record AlunoDashboardResponse(
        SaudacaoResponse saudacao,
        PeriodoVigenteResponse periodoVigente,
        Boolean alertaPeriodoAusente,
        KpisResponse kpis,
        List<PendenciaResponse> pendencias,
        List<UltimaSolicitacaoResponse> ultimasSolicitacoes,
        List<ProximoEventoResponse> proximosEventos,
        @JsonProperty("_links") Map<String, String> links
) {

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record SaudacaoResponse(String nome, String curso) {
    }

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record PeriodoVigenteResponse(UUID id, int ano, int semestre, String rotulo, LocalDate inicio, LocalDate fim) {
    }

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record KpisResponse(
            KpiHorasFormativas horasFormativas,
            Integer solicitacoesAbertas,
            Integer eventosHoje,
            Integer certificados
    ) {
    }

    public record KpiHorasFormativas(int validadas, int requeridas) {
    }

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record PendenciaResponse(UUID id, String titulo, String estado, String href) {
    }

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record UltimaSolicitacaoResponse(
            UUID id,
            String protocolo,
            String tipoNome,
            String estado,
            OffsetDateTime prazoEm,
            boolean slaVencido
    ) {
    }

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public record ProximoEventoResponse(
            UUID id,
            String titulo,
            OffsetDateTime inicioEm,
            OffsetDateTime fimEm,
            boolean janelaAtiva,
            String href
    ) {
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/bff/api/dto/AlunoDashboardResponse.kt
package br.ufpr.sept.so2.modules.bff.api.dto

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.LocalDate
import java.time.OffsetDateTime
import java.util.UUID

@JsonInclude(JsonInclude.Include.ALWAYS)
data class AlunoDashboardResponse(
    val saudacao: SaudacaoResponse,
    val periodoVigente: PeriodoVigenteResponse?,
    val alertaPeriodoAusente: Boolean?,
    val kpis: KpisResponse,
    val pendencias: List<PendenciaResponse>?,
    val ultimasSolicitacoes: List<UltimaSolicitacaoResponse>?,
    val proximosEventos: List<ProximoEventoResponse>?,
    val pendenciasFormativas: List<PendenciaResponse>?,
    @get:JsonProperty("_links")
    val links: Map<String, String>,
) {
    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class SaudacaoResponse(val nome: String, val curso: String?)

    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class PeriodoVigenteResponse(
        val id: UUID,
        val ano: Int,
        val semestre: Int,
        val rotulo: String,
        val inicio: LocalDate,
        val fim: LocalDate,
    )

    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class KpisResponse(
        val horasFormativas: KpiHorasFormativas?,
        val solicitacoesAbertas: Int?,
        val eventosHoje: Int?,
        val certificados: Int?,
    )

    data class KpiHorasFormativas(val validadas: Int, val requeridas: Int)

    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class PendenciaResponse(
        val id: UUID,
        val titulo: String,
        val estado: String,
        val href: String,
    )

    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class UltimaSolicitacaoResponse(
        val id: UUID,
        val protocolo: String,
        val tipoNome: String,
        val estado: String,
        val prazoEm: OffsetDateTime?,
        val slaVencido: Boolean,
    )

    @JsonInclude(JsonInclude.Include.ALWAYS)
    data class ProximoEventoResponse(
        val id: UUID,
        val titulo: String,
        val inicioEm: OffsetDateTime,
        val fimEm: OffsetDateTime,
        val janelaAtiva: Boolean,
        val href: String,
    )
}
